### PR TITLE
[ATen] Support multi dim any and all reductions

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110311
* __->__ #110310

This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

### CODE REVIEW

File: test/onnx/test_fx_op_consistency.py [hunk 1]
```diff
@@ -57,11 +57,11 @@
         "acosh",
         "add",
         "addmm",
-        "all",
+        # "all",    - onnxscript doesn't handle aten::all.dims
         "allclose",
         "amax",
         "amin",
-        "any",
+        # "any",    - onnxscript doesn't handle aten::any.dims
```
Comment (peterbell10) [inline test/onnx/test_fx_op_consistency.py:64]: Apparently this is testing out-of-tree code, so can't be fixed in this PR
https://github.com/microsoft/onnxscript/blob/f8046e17490a3222d4ee4bca346b4449a2d30ae4/onnxscript/function_libs/torch_lib/ops/core.py#L404

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -1541,10 +1552,46 @@ TORCH_IMPL_FUNC(any_out)
...
+
 TORCH_IMPL_FUNC(any_all_out)(const Tensor& self, const Tensor& result) {
   allany_impl<0>(self, result, {}, false, or_stub);
 }
 
+template <bool is_all>
+Tensor allany_dims_default(const Tensor &self, OptionalIntArrayRef dim, bool keepdim) {
+  // Default implementation in terms of all-reduce or single dim reduce
+  if (!dim) {
+    auto out = is_all ? self.all() : self.any();
```
Comment (lezcano) [inline aten/src/ATen/native/ReduceOps.cpp:1568]: if constexpr, although the compiler should be able to optimise out the branch.

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py [hunk 3]
```diff
@@ -8,17 +8,14 @@
 import parameterized
 
 import torch
+from onnx_test_common import MAX_ONNX_OPSET_VERSION, MIN_ONNX_OPSET_VERSION
```
Comment (lezcano) [inline test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:11]: why these changes?

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -295,6 +295,9 @@
 - name: any.dim(Tensor self, int dim, bool keepdim=False) -> Tensor
   output_differentiability: [False]
 
+- name: any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
+  output_differentiability: [False]
+
```
Comment (lezcano) [inline tools/autograd/derivatives.yaml:300]: Can't you make it composite implicit, even if it's not differentiable?

File: tools/pyi/gen_pyi.py [hunk 5]
```diff
@@ -171,6 +171,12 @@ def should_bind_method(python_func: PythonSignatureNativeFunctionPair) -> bool:
     "copy_",
 ]
 
+overload_blocklist = [
+    # Overloads which are more specific than another, so not needed
+    "all.dim",
```
Comment (lezcano) [inline tools/pyi/gen_pyi.py:176]: nit. Put the name of the other overload.

File: torch/_refs/__init__.py [hunk 6]
```diff
@@ -2257,7 +2249,10 @@ def any(
     keepdim: bool = False,
 ) -> TensorLikeType:
     a_ = _maybe_convert_to_dtype(a, torch.bool)
-    result = ne(sum(a_, dim=dim, keepdim=keepdim), False)  # type: ignore[arg-type]
+    if isinstance(dim, (list, tuple)) and len(dim) == 0:
+        result = a_.clone()
+    else:
+        result = ne(sum(a_, dim=dim, keepdim=keepdim), False)  # type: ignore[arg-type]
```
Comment (lezcano) [inline torch/_refs/__init__.py:2255]: Preexisting, but we should probably be calling `torch.`  ops

File: aten/src/ATen/native/ReduceOps.cpp [hunk 7]
```diff
@@ -1541,10 +1552,46 @@ TORCH_IMPL_FUNC(any_out)
...
+    if (keepdim) {
+      DimVector out_shape(self.dim(), 1);
+      return out.expand(out_shape);
+    }
+    return out;
+  }
+
+  Tensor out = self;
+  for (auto d : *dim) {
```
Comment (lezcano) [inline aten/src/ATen/native/ReduceOps.cpp:1577]: This is fine, but couldn't we extend the tensoriterator implementation? I reckon it should be straightforward and it should be much more efficient.

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py (hunk 3 above)
Comment (peterbell10) [reply test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:11]: Those constants aren't defined in the file they were being imported from. This file must not be run in CI.

File: tools/autograd/derivatives.yaml (hunk 4 above)
Comment (peterbell10) [reply tools/autograd/derivatives.yaml:300]: No, because that would have higher priority than the `CPU` and `CUDA` dispatch.

File: tools/pyi/gen_pyi.py (hunk 5 above)
Comment (peterbell10) [reply tools/pyi/gen_pyi.py:176]: wdym? If we skip the array overload then mypy would complain when you pass a list of dims.

File: aten/src/ATen/native/ReduceOps.cpp (hunk 7 above)
Comment (peterbell10) [reply aten/src/ATen/native/ReduceOps.cpp:1577]: CPU and CUDA do use tensoriterator. I included this mainly for MPS or any third-party dispatch keys.

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py (hunk 3 above)
Comment (lezcano) [reply test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:11]: mind filing an issue?

File: aten/src/ATen/native/native_functions.yaml [hunk 8]
```diff
@@ -681,15 +681,29 @@
...
   dispatch:
     CPU, CUDA: all_out
     MPS: all_out_mps
 
+- func: all.dims_out(Tensor self, int[1]? dim=None, bool keepdim=False, *, Tensor(a!) out) -> Tensor(a!)
+  device_check: NoCheck   # TensorIterator
+  structured: True
+  dispatch:
+    CPU, CUDA: all_dims_out
+    MPS: all_dims_out_mps
```
Comment (lezcano) [inline aten/src/ATen/native/native_functions.yaml:704]: does mps implement all but not any?

Comment (peterbell10) [reply aten/src/ATen/native/native_functions.yaml]: Oh sorry that's dead code. I looked briefly into adding it but the MPS code was fairly complex and I have no way to test so I decided against it. Removed in the latest commit.

**Review by lezcano — APPROVED** [review]

File: test/onnx/test_fx_op_consistency.py (hunk 1 above)
Comment (peterbell10) [reply test/onnx/test_fx_op_consistency.py:64]: cc @justinchuby

Comment (justinchuby) [reply test/onnx/test_fx_op_consistency.py:64]: The current best thing to do is to xfail it here: https://github.com/pytorch/pytorch/pull/110310/files/73dd8600a8097f98b675510024e01ea1a3d6af5d#diff-db2f78a51511bb172cbfde1b2f68272b8b33049abe2571cded27bcd0f3ae5fa4R176 and track it with an issue. 

I will discuss with the team to find a better solution forward. Thanks!

File: aten/src/ATen/native/native_functions.yaml [hunk 9]
```diff
@@ -681,15 +681,29 @@
   structured_delegate: all.out
   variants: function, method
 
+- func: all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```
Comment (justinchuby) [inline aten/src/ATen/native/native_functions.yaml:684]: Question: Does `int[1]` imply the length of the sequence can only be 1? I have seen the use of `int[]` too. Was this intended?

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py (hunk 3 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:11]: Thanks!

**Comment by justinchuby** [comment]
When `dim` is None and `keepdims` is False for `all.dims`, it is equivalent to `all.default`?

File: aten/src/ATen/native/native_functions.yaml (hunk 9 above)
Comment (peterbell10) [reply aten/src/ATen/native/native_functions.yaml]: `int[1]` is a broadcasting list, it means if you pass a scalar it will be broadcast to a list of size 1. I'm removing the broadcasting aspect though as it's covered by the `all.dim` overload (I had hoped  we wouldn't need two).

**Comment by peterbell10** [comment]
> When dim is None and keepdim is False for all.dims, it is equivalent to all.default? Just curious about when dim needs to take None. Would the alternative be adding keepdim=False to all.default? I assume there was a reason not to do that?

Yes, `dim=None` is an all-reduce. The reason to prefer this signature is for generic code where you accept a `dim` argument from the caller, if we can't pass `dim=None` then you need to write code like:
```python
return x.all() if dim is None else x.all(dim=dim)
```

**Comment by justinchuby** [comment]
> > When dim is None and keepdim is False for all.dims, it is equivalent to all.default? Just curious about when dim needs to take None. Would the alternative be adding keepdim=False to all.default? I assume there was a reason not to do that?
> 
> Yes, `dim=None` is an all-reduce. The reason to prefer this signature is for generic code where you accept a `dim` argument from the caller, if we can't pass `dim=None` then you need to write code like:
> 
> ```python
> return x.all() if dim is None else x.all(dim=dim)
> ```

**Comment by justinchuby** [comment]
Sorry wrong click

File: test/onnx/test_fx_op_consistency.py [hunk 10]
```diff
@@ -188,6 +188,9 @@
         "addmm", dtypes=onnx_test_common.BOOL_TYPES,
         reason=onnx_test_common.reason_onnx_does_not_support("Addmm")
     ),
+    xfail(
+        "all", reason=onnx_test_common.reason_onnx_script_does_not_support("aten.all.dims")
```
Comment (thiagocrepaldi) [inline test/onnx/test_fx_op_consistency.py:192]:
```suggestion
        "all", reason=onnx_test_common.reason_onnx_script_does_not_support("aten.all.dims") + "https://github.com/pytorch/pytorch/issues/XXXXX"
```

Maybe add the tracking issue as part of the reason or a `# TODO: Remove after https://github.com/pytorch/pytorch/issues/XXXXX` comment around it?

@justinchuby how do you feel changing `this to def reason_onnx_script_does_not_support(operator, dtypes, github_issue=None)` to help tracking the underlying issue. That is a way to remind users to create issues without our request

Comment (justinchuby) [reply test/onnx/test_fx_op_consistency.py:192]: I will update https://github.com/pytorch/pytorch/pull/110536

File: test/onnx/test_fx_op_consistency.py [hunk 11]
```diff
@@ -188,6 +188,9 @@
         "addmm", dtypes=onnx_test_common.BOOL_TYPES,
         reason=onnx_test_common.reason_onnx_does_not_support("Addmm")
     ),
+    xfail(
```
Comment (justinchuby) [inline test/onnx/test_fx_op_consistency.py:191]:
```suggestion
    xfail_torchlib_forward_compatibility(
```

Could you rebase on a commit after https://github.com/pytorch/pytorch/pull/110536 and use `xfail_torchlib_forward_compatibility`? Thanks!

File: test/onnx/test_fx_op_consistency.py [hunk 12]
```diff
@@ -201,6 +204,9 @@
         "amin", dtypes=(torch.int16, *onnx_test_common.BOOL_TYPES),
         reason=onnx_test_common.reason_dynamo_does_not_support("ReduceMin", "bool, int16")
     ),
+    xfail(
```
Comment (justinchuby) [inline test/onnx/test_fx_op_consistency.py:207]:
```suggestion
    xfail_torchlib_forward_compatibility(
```

**Comment by peterbell10** [comment]
@albanD do you have any idea about these segfaults in MPS? I tried disabling `CPU, CUDA` dispatch locally so I was testing with the `CompositeExplicitAutograd` version on CPU and it worked fine, so there seems to be an MPS-specific issue here.

**Comment by peterbell10** [comment]
@kulinseth ping

**Comment by kulinseth** [comment]
> cc @kulinseth any idea how we can be segfaulting here? :o

It could be related to https://github.com/pytorch/pytorch/issues/95538. For >4D reduction ops are crashing.

**Comment by peterbell10** [comment]
Okay if it's a pre-existing issue then I'll just add a skip.

**Comment by albanD** [comment]
If the error is a hard crash, you will have to add it to the skip list instead of xfail.

**Review by albanD — APPROVED** [review]
SGTM

**Review by justinchuby — APPROVED** [review]
